tensorflow_addons/image/filters.py

```diff
@@ -204,3 +200,95 @@ def median_filter2d(
+    constant_values: TensorLike = 0,
+    name: Optional[str] = None,
+) -> FloatTensorLike:
+    """
+    This function is responsible for having Gaussian Blur. It takes the image as input, computes a gaussian-kernel
+    which follows normal distribution then convolves the image with the kernel. It is implemented as 2 1D convolutions.
+    Args:
+    image: A tensor of shape
+        (batch_size, height, width, channels)
+        (NHWC), (batch_size, channels, height, width)(NCHW).
+
+    sigma:A constant of type float32. It is the standard deviation of the normal distribution.
+          The more the sigma, the more the blurring effect.
+          G(x,y)=1/(2*3.14*sigma**2)e^((x**2+y**2)/2sigma**2)
+          In 1D,
+          G(x)=e^(-x**2)/2*sigma**2
+
+    filter_shape:An `integer` or `tuple`/`list` of 2 integers, specifying
+        the height and width of the 2-D gaussian filter. Can be a single integer
+        to specify the same value for all spatial dimensions.
+
+    padding:A `string`, one of "REFLECT", "CONSTANT", or "SYMMETRIC".
+        The type of padding algorithm to use, which is compatible with
+        `mode` argument in `tf.pad`. For more details, please refer to
+        https://www.tensorflow.org/api_docs/python/tf/pad.
+
+    constant_values:A `scalar`, the pad value to use in "CONSTANT"
+        padding mode.
+    name: A name for this operation (optional).
+    Returns:
+        3D or 4D 'Tensor' of same type float64.
+    Raises:
+        Value error if:
+            1). Sigma=0
+            2). passing some string other than ["REFLECT", "CONSTANT", "SYMMETRIC"] in padding.
+    """
+
+    with tf.name_scope(name or "gaussian_filter2d"):
+        if sigma == 0:
+            raise ValueError("Sigma should not be zero")
+        if padding not in ["REFLECT", "CONSTANT", "SYMMETRIC"]:
+            raise ValueError("Padding should be REFLECT, CONSTANT, OR SYMMETRIC")
+
+        image = tf.cast(image, tf.float32)
+        image = img_utils.to_4D_image(image)
+        channels = tf.shape(image)[3]
+        filter_shape = keras_utils.normalize_tuple(filter_shape, 2, "filter_shape")
+
+        gaussian_filter_x = _get_gaussian_kernel(sigma, filter_shape[1], channels)
+        gaussian_filter_x = tf.repeat(gaussian_filter_x, channels)
+        gaussian_filter_x = tf.reshape(
+            gaussian_filter_x, [1, filter_shape[1], channels, 1]
+        )
+
+        gaussian_filter_x = tf.cast(gaussian_filter_x, tf.float32)
+        gaussian_filter_y = _get_gaussian_kernel(sigma, filter_shape[0], channels)
+        gaussian_filter_y = tf.repeat(gaussian_filter_y, channels)
+        gaussian_filter_y = tf.reshape(
+            gaussian_filter_y, [filter_shape[0], 1, channels, 1]
+        )
+
+        gaussian_filter_y = tf.cast(gaussian_filter_y, tf.float32)
+        image = _pad(
+            image, filter_shape, mode=padding, constant_values=constant_values,
+        )
+        conv_ops_x = tf.nn.depthwise_conv2d(
```

Comment:
Are we sure that apply separable filters it is faster down the stack?
Cause I see Google internals use a single call https://github.com/google/automl/blob/master/efficientdet/aug/autoaugment.py#L1127

Reply:
@bhack  The link in which you have shown uses a very small kernel(3*3). So there it is no effect on the speed(therefore complexity to program is seen).
I have seen other libraries-https://github.com/scipy/scipy/blob/v1.4.1/scipy/ndimage/filters.py#L220-L303
They use multidimentional convolution as sequence of 1d convolution.
The 2d direct convolution requires more multiplication than sequence of 1d convolution.
Consider a (3*3) matrix and a (3*3) kernel. The regular 2d convolution requires 9 multiplication and 9 addition per sample. The sequence of 1d convolutions is like this-6 multiplication(3 for x and y direction each) and 4 additions. Clearly, the difference would increase if kernel_size increases. 
I allowed the user to pass the kernel size which may be large.

Reply:
Are you sure that are not separated internally?

Reply:
All 2d convolution_filters are not separable(and the depthwise_conv2d is made keeping in mind the generality).
The gaussian filter is separable(since e^(x^2+y^2)=e^(x^2).e^(y^2)).
I have gone through the code and documentation. It implements as normal(to cope with non separable convolution_filters).

Reply:
Yes separation is not performed internally. As all conv_filters are not separable.
For gaussian filter(e**((x**2)+(y**2))=e(x**2).e**(y**2)). Thus separable convolution filters can also be made.

Reply:
Yes I meant that it just seemed strange that we didn't have a flag/arg to tell that the filter are separabile for handling this internally. Also have we any significant overhead with smaller kernels with two API calls?

Reply:
Can you please elaborate the last point?

Reply:
I meant if for quite pratical kernel i.e. <= 7x7 it create less overhead a single call with a 2d filter.

Reply:
@bhack The gaussian filter is convolved with an image(which generally have large dimensions). So IMO there would be effect of applying simple gaussian kernel and sequence of 2 1d kernels. Besides we dont know in which situations, the gaussian kernel is used. The user may put large values of the kernel_size. Besides all other libraries-scipy,opencv,skimage uses this technique.
Also computing 2d kernels will cost the python code two for loops(presently all things are handled internally by tensorflow) which will slow down the speed.

Reply:
Frankly, no offense, not just to **say** the performance but to **measure** the performance. The overhead of launching tensorflow kernel is not negligible.

Reply:
@WindQAQ can you elaborate? I cannot get you.

Reply:
As far as I can understand you are wanting mathematical performance.
consider the image of size M * N {1024 * 768 (Grayscale)} and kernel =(k * k)(7 * 7)
Computational complexity: Convolution -> O(MNkk)
Computational complexity: Separable convolution -> O(2∗MNk)
Using normal convolution you got O(1024*768*7*7) = 38535168 operations.
Using separable convolution you got(2*1024*768*7) = 11010048 operations.
I m sorry if i'm mistaken to understand you. Please clear if I am mistaken.

Reply:
@ghosalsattam I think that generally It could be fast enough to handle the two kernel launch in the most common used image kernel size.
What @WindQAQ is talking about is https://www.tensorflow.org/guide/profiler (See Performance summary point)

Reply:
I mean it's better to see some numeric results that compare two 1D separable filters with single 2D filter. I know that separable filters  approach are theoretically faster than single filter,. However, because it involves GPU kernel (the call to `depthwise_conv2d`), things might be different. A single thread taking `n x n` operations together is not necessarily slower than single thread taking `n` operations twice consecutively in parallel computing when it comes to the memory coherence and shared memory etc.

https://www.tensorflow.org/guide/profiler
https://www.hpcs.cs.tsukuba.ac.jp/icpp2019/data/posters/Poster17-abst.pdf